collect_app/src/main/java/org/odk/collect/android/utilities/TextUtils.java

```diff
@@ -69,28 +69,35 @@ private TextUtils() {
 
     }
 
-    protected static String markdownToHtml(String text) {
+    public static String markdownToHtml(String text) {
 
-        text = text.replaceAll("<([^a-zA-Z/])", "&lt;$1");
+        text = text.replaceAll("<([^a-zA-Z/])", "&gm;$1");
         // https://github.com/enketo/enketo-transformer/blob/master/src/markdown.js
 
         // span - replaced &lt; and &gt; with <>
-        text = ReplaceCallback.replace("(?s)<\\s?span([^\\/\n]*)>((?:(?!<\\/).)+)<\\/\\s?span\\s?>",
+        text = ReplaceCallback.replace("/&lt;\\s?span([^/\\n]*)&gt;((?:(?!&lt;\\/).)+)&lt;\\/\\s?span\\s?&gt;/gm _createSpan >",
                 text, createSpan);
         // strong
-        text = text.replaceAll("(?s)__(.*?)__", "<strong>$1</strong>");
-        text = text.replaceAll("(?s)\\*\\*(.*?)\\*\\*", "<strong>$1</strong>");
+        text = text.replaceAll("/__(.*?)__/gm", "<strong>$1</strong>");
+        text = text.replaceAll("/\\*\\*(.*?)\\*\\*/gm", "<strong>$1</strong>");
         // emphasis
-        text = text.replaceAll("(?s)_([^\\s][^_\n]*)_", "<em>$1</em>");
-        text = text.replaceAll("(?s)\\*([^\\s][^\\*\n]*)\\*", "<em>$1</em>");
+        text = text.replaceAll("/_([^\\s][^_\\n]*)_/gm", "<em>$1</em>");
+        text = text.replaceAll("/\\*([^\\s][^*\\n]*)\\*/gm", "<em>$1</em>");
         // links
-        text = text.replaceAll("(?s)\\[([^\\]]*)\\]\\(([^\\)]+)\\)",
+        text = text.replaceAll("/\\[([^\\]]*)\\]\\(([^)]+)\\)/gm",
                 "<a href=\"$2\" target=\"_blank\">$1</a>");
         // headers - requires ^ or breaks <font color="#f58a1f">color</font>
         text = ReplaceCallback.replace("(?s)^(#+)([^\n]*)$", text, createHeader);
         // paragraphs
         text = ReplaceCallback.replace("(?s)([^\n]+)\n", text, createParagraph);
 
+        // "\" will be used as escape character for *, _
+        text = text
+                .replaceAll(" /&/gm", "'&amp;'")
+                .replaceAll("/\\\\/gm", "'&92;'")
+                .replaceAll(" /\\\\*/gm", "'&42;'")
```

Comment:
This doesn’t look right — a single backslash should be use for the escape. Can you please add tests for this and the two escape sequences above?

Reply:
Hello @lognaturel using http://www.vogella.com/tutorials/JavaRegularExpressions/article.html#backslashes-in-java has a proof(number 3.7. Backslashes in Java to be specific) In Java when you want to add a single backslash \ it is done with four \\\\ backslashes. So the codes above actually mean .replace("/\/gm", "'&92'"). I have added tests to proof this for the last and second to the last code above, made a commit and pushed to Git. 
But for the first line of codes is there for the theoretical case that the original string contains e.g. "&92;" or "&42;" or another internally used escape character. These strings would be converted in the subsequent lines when they shouldn't be. Martijn explained it better here https://github.com/enketo/enketo-transformer/commit/52c60b7ad20da035661ead3ea782480ff6e40ede
Please review